src/arch/x86_64.rs

```diff
@@ -71,19 +71,28 @@ impl BootInfo for RustbootBootInfo {
             return None;
         }
 
-        let (cfg, buf) = unsafe {
-            // Safety: we can reasonably assume this will only be called
-            // after `arch_entry`, so if we've failed to initialize the
-            // framebuffer...things have gone horribly wrong...
-            FRAMEBUFFER.get_unchecked()
-        };
-        Some(Framebuffer::new(cfg, buf.lock()))
+        Some(unsafe { Self::mk_framebuf() })
     }
 
     fn subscriber(&self) -> Option<tracing::Dispatch> {
-        Some(tracing::Dispatch::new(
-            hal_x86_64::tracing::Subscriber::default(),
-        ))
+        use mycelium_trace::{writer::MakeWriterExt, Subscriber};
+        if !self.has_framebuffer {
+            // TODO(eliza): we should probably write to just the serial port if
+            // there's no framebuffer...
```

Comment:
That does seem like a good idea. Perhaps follow-up work to make the framebuf optional in the subscriber?

Reply:
yeah, i was gonna do that in a subsequent branch